src/resilience/backend/FenixBackend.cpp

```diff
@@ -0,0 +1,389 @@
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+ */
+#include "FenixBackend.hpp"
+
+#include <sstream>
+#include <fstream>
+#include <cassert>
+
+#include <fenix.h>
+
+#include <resilience/AutomaticCheckpoint.hpp>
+
+#ifdef KR_ENABLE_TRACING
+#include <resilience/util/Trace.hpp>
+#endif
+
+#include <fenix.h>
+
+// TODO: implement exception class
+//Kokkos::Impl::throw_runtime_exception( out.str() );
+#define FENIX_THROW( msg ) std::cerr << msg << std::endl
+
+namespace KokkosResilience
+{
+  class FenixClient
+  {
+    using serializer_t   = std::function< bool( std::ostream & ) >;
+    using deserializer_t = std::function< bool( std::istream & ) >;
+
+    struct ProtectedMemoryBlock
+    {
+      explicit ProtectedMemoryBlock( int id, const serializer_t &serializer, const deserializer_t &deserializer )
+          : m_id( id ), m_serializer( serializer ), m_deserializer( deserializer ) {}
+
+      int m_id;
+      serializer_t m_serializer;
+      deserializer_t m_deserializer;
+      std::vector<char> m_buffer;
+      std::size_t m_size;
+      bool m_protect = false;
+      bool m_registered = false;
+    };
+    
+  public:
+    // TODO Enable user configuration options, especially in regards to data group policy and checkpoint depth
+    FenixClient( MPI_Comm comm ) : m_comm( comm )
+    {
+      create_data_group();
+    }
+
+    ~FenixClient()
+    {
+      clear_registry();
+      delete_data_group();
+    }
+
+    void protect_member( int hash, serializer_t serializer, deserializer_t deserializer )
+    {
+      int member_id = g_current_member_id++;
+
+      auto item = m_registry.find( hash );
+      if ( item == m_registry.end() )
+      {
+        item = m_registry.emplace( std::piecewise_construct,
+                                   std::forward_as_tuple( hash ),
+                                   std::forward_as_tuple( member_id, serializer, deserializer ) ).first;
+      }
+
+      item->second.m_protect = true;
+    }
+
+    void unprotect_member( int hash )
+    {
+      auto item = m_registry.find( hash );
+      if ( item != m_registry.end() && item->second.m_registered )
+      {
+        Fenix_Data_member_delete( m_group_id, item->second.m_id );
+        item->second.m_protect = false;
+        item->second.m_registered = false;
+      }
+      m_registry.erase( item );
+    }
+
+    void checkpoint( const std::unordered_set< int > &hashes, int version )
+    {
+      for ( auto &&hash : hashes )
+      {
+        auto item = m_registry.find( hash );
+        if ( item != m_registry.end() )
+        {
+          // check if we need to create new Fenix data member
+          if (item->second.m_protect && !item->second.m_registered)
+          {
+            // load serialized object into stream
+            std::ostringstream stream;
```

Comment:
I'm not 100% sure of using ostringstream here. I think we should discuss this more at the meeting as we may want some extra support for in-memory checkpointing "memory streams"